### PR TITLE
Ground Truth Authentication

### PR DESCRIPTION
**Pull request by Nicholas714**
- [x] Button that authenticates the user using Ground Truth
- [x] Button that allows a user to log out
- [x] Automatic login after first sign in
- [x] Store client secret securely (using PKCE)
- [x] Pull data on app start and provide as context to other components 
- [x] Working on iOS
- [x] Working on Android

Creating a login/logout button with the whole auth flow would look like this:
```
<AuthContext.Consumer>
    {({ user, login, logout }) => {
        <Button onPress={user ? logout : login} title={user ? `Logout ${user.name}` : "Login"}/>
     }}
</AuthContext.Consumer>
```
where `user` will have login details (see screenshot), `login` is a function that starts login flow and `logout` will delete user contents and log them out of GroundTruth

image
image

### CODE REVIEW

**Comment by evan10s** [comment]
I _may_ be out of the loop but is there a / can we fix the "Authorizing will redirect to *null*" on the Ground Truth approval screen?

**Comment by Nicholas714** [comment]
> I _may_ be out of the loop but is there a / can we fix the "Authorizing will redirect to _null_" on the Ground Truth approval screen?

I think this is because the redirect url is not an actual link, its gt.hack.login://redirect. The library we are using forces it to not be an actual link

**Comment by joel99** [comment]
If that's the case can a change go in on ground truth to not show that line if null? @petschekr or not sure if there's a better behavior

**Comment by petschekr** [comment]
Whoops didn't expect weird URLs like that when formatting them
Will submit a bugfix PR

File: App.js [hunk 1]
```diff
@@ -1,69 +1,182 @@
...
+    faList,
+    faBell,
+    faKey,
+    faQuestion
+} from "@fortawesome/free-solid-svg-icons";
+import AsyncStorage from "@react-native-community/async-storage";
+import { authorize } from "react-native-app-auth";
+
+export const AuthContext = React.createContext();
+const authUrl = "https://login.hack.gt";
```
Comment (joel99) [inline App.js]: Seems a bit tedious, make a TODO/open an issue to move to env var
https://stackoverflow.com/questions/33117227/setting-environment-variable-in-react-native

File: App.js [hunk 2]
```diff
@@ -1,69 +1,182 @@
...
+
+const config = {
+    clientId:
+        "7d1c11b30351e91d6517492c19a9a0185a6e4e6304f9826f96a76895534cf26f",
+    redirectUrl: "gt.hack.live://redirect",
+    clientSecret: "hackgt",
```
Comment (joel99) [inline App.js]: so I only skimmed pkce, is this your code challenge...

File: App.js [hunk 3]
```diff
@@ -1,69 +1,182 @@
...
+    }
 
-  render() {
-    return (
-      <AppContainer>Hi</AppContainer>
-    )
-  }
-}
\ No newline at end of file
+    logout = async () => {
```
Comment (joel99) [inline App.js]: for consistency with react methods, would prefer you make these function declarations over expressions (`function logout() {`). Won't enforce. Difference is not too big

File: App.js [hunk 4]
```diff
@@ -1,69 +1,182 @@
...
+        try {
+            const result = await authorize(config);
+
+            this.setState({ accessToken: result.accessToken });
+
+            this.fetchUserDetails(result.accessToken);
+
+            AsyncStorage.setItem("accessToken", accessToken);
+        } catch (error) {
+            console.log(error);
```
Comment (joel99) [inline App.js]: Open an issue and log a todo comment about surfacing errors to UI

File: ios/AppGT.xcworkspace/xcshareddata/IDEWorkspaceChecks.plist [hunk 5]
```diff
@@ -0,0 +1,8 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (joel99) [inline ios/AppGT.xcworkspace/xcshareddata/IDEWorkspaceChecks.plist]: should this be checked in?

File: App.js (hunk 4 above)
Comment (joel99) [reply App.js]: I think also mention in the issue - surface login success greeting/land user on some page that indicates login success

File: App.js (hunk 2 above)
Comment (Nicholas714) [reply App.js]: Nope code challenge is automatically generated.

File: ios/AppGT.xcworkspace/xcshareddata/IDEWorkspaceChecks.plist (hunk 5 above)
Comment (Nicholas714) [reply ios/AppGT.xcworkspace/xcshareddata/IDEWorkspaceChecks.plist]: yep I think so

**Pull request merged**